// @lableb/javascript-sdk/src/http/browser/index.ts
import { httpDelete } from "./delete";
import { httpGet } from "./get";
import { httpPost } from "./post";
import { httpPostFormData } from "./post-form-data";
import { httpPut } from "./put";
import { httpPutFormData } from "./put-form-data";



export interface HttpClientAPI {
    get: typeof httpGet,
    post: typeof httpPost,
    postFormData: typeof httpPostFormData,
    delete: typeof httpDelete,
    put: typeof httpPut,
    putFormData: typeof httpPutFormData,
}



/**
 * Browser http client that use `fetch` internally to make the requests
 */
export function BrowserHttpClient(): HttpClientAPI {

    const API = {

        /** usual get request */
        get: httpGet,

        /** usual post request that use json for body */
        post: httpPost,

        /** post request that use form-data body type */
        postFormData: httpPostFormData,

        /** usual delete request */
        delete: httpDelete,

        /** usual put request */
        put: httpPut,

        /** put request that use form-data body type */
        putFormData: httpPutFormData,
    }

    return API;
}
