// @lableb/javascript-sdk/src/core/autocomplete-feedback.ts
import { LablebSDKContext } from ".";
import { AutocompleteFeedbackURLOptions } from "../types/url";


/**
 * Send autocomplete feedback to Lableb Cloud API
 * 
 * @param {object} options 
 * @param {string} options.query
 * @param {string} options.itemURLSchema
 * @param {number} options.itemId
 * @param {number} options.itemOrder
 * @param {string} options.url
 * 
 * @returns {Promise} - feed back response
 */
export function autocompleteFeedback(

    this: LablebSDKContext,

    options: Omit<AutocompleteFeedbackURLOptions, "sessionId">

) {


    return this.httpClient.get({
        url: this.urlBuilder.autocompleteFeedbackURL({
            ...options,
            sessionId: this.sessionId,
        })
    });
}






/**
 * Send autocomplete feedback to Lableb Cloud API from a given URL
 * 
 * @param {string} url - request url
 * 
 * @returns {Promise} - feed back response
 */
export function autocompleteFeedbackFromURL(

    this: LablebSDKContext,

    url: string

) {

    return this.httpClient.get({
        url,
    });
}
