// @lableb/javascript-sdk/src/core/indexing.ts
import { LablebDocument } from "../types/sdk/lableb-document";
import { LablebSDKContext } from "./index";


/**
 * Index documents at Lableb Cloud
 * pass array of objects to index(or update) them immediately 
 * 
 * @param {object[]} documents
 * 
 * @returns {Promise} - index document response
 */
export function indexDocuments(

    this: LablebSDKContext,

    documents: LablebDocument[]

) {
    /**
     * you cannot use the delete function when the build target is commonJs
     * and the sdk is used in react/angular application
     * 
     * we need to write the full function code here, because writing it in separate function
     * and calling it will have `this` bound to SDK context and not the global object
     */
    if (new Function("try { return this === window; } catch(e){ return false; }")()) {
        throw new Error('You cannot use index function in browser');
    }

    return this.httpClient.post({
        url: this.urlBuilder.indexDocumentURL(),
        body: documents,
    });
}
