// @lableb/javascript-sdk/src/core/recommend-feedback.ts
import { LablebSDKContext } from ".";
import { RecommendFeedbackURLOptions } from "../types/url";


/**
 * Send Recommend Feedback to Lableb
 * 
 * @param {object} options 
 * @param {number} options.sourceId - recommendation input document id (that was responsible for generating the recommendations)
 * @param {string} options.sourceTitle - recommendation input document title
 * @param {string} options.sourceUrl - recommendation input document url
 * @param {number} options.targetId - user chosen recommendation item id (that was returned from the recommendation response)
 * @param {string} options.targetTitle - user chosen recommendation item title
 * @param {string} options.targetUrl - user chosen recommendation item url
 * @param {number} options.itemOrder - user chosen recommendation item order in recommendation response
 * @param {string} options.sessionId - unique session id per user
 * @param {number} options.userId - user id
 * @param {string} options.userIp - user ip address
 * @param {string} options.userCountry - iso 3166 country code
 *
 * @returns {Promise}
 */
export function recommendFeedback(

    this: LablebSDKContext,

    options: Omit<RecommendFeedbackURLOptions, "sessionId">

) {

    return this.httpClient.get({
        url: this.urlBuilder.recommendFeedbackURL({
            ...options,
            sessionId: this.sessionId,
        }),
    });
}



/**
 * Send Recommend Feedback to Lableb from a given url
 * 
 * @param url 
 * 
 * @returns {Promise} - recommend response
 */
export function recommendFeedbackFromURL(

    this: LablebSDKContext,

    url: string

) {
    return this.httpClient.get({
        url,
    });
}
