// @lableb/javascript-sdk/src/url/indexing.ts
import { customIdentity } from "../helpers";
import { pickBy } from "../helpers/pick-by";
import { URLBuilderExecutionContext } from "../types/url";
import { buildURLParams } from "./url.utils";



/**
 * Generate Index Documents URL
 * 
 * it use the options passed to the SDK function to generate the index document request's url
 * 
 * @return {string} - URL used for the Index Document request
 */
export function indexDocumentURL(this: URLBuilderExecutionContext) {

    /**
     * check properties existence in options
     */
    this.checkExistence({
        functionName: 'indexDocumentURL',
        properties: ["projectName", "collectionName", "indexToken"]
    });


    /**
     * destructor the properties
     */
    const { baseURL, projectName, collectionName, indexToken } = this.options;

    /**
     * remove undefined values and build the url
     */
    const urlParams = buildURLParams(
        pickBy(
            {
                token: indexToken,
            },
            customIdentity
        )
    );

    return `${baseURL}/projects/${projectName}/collections/${collectionName}/documents?${urlParams}`;
}









/**
 * Generate Update Documents URL
 * 
 * @return {string} URL used for the Update Document request
 */
export function updateDocumentURL(this: URLBuilderExecutionContext) {

    this.checkExistence({
        functionName: 'updateDocumentURL',
        properties: ["projectName", "collectionName", "indexToken"]
    });

    return this.indexDocumentURL();
}
