// @lableb/javascript-sdk/src/types/url/url.ts

import { autocompleteURL } from "../../url/autocomplete";
import { autocompleteFeedbackURL } from "../../url/autocomplete-feedback";
import { deleteDocumentURL } from "../../url/delete";
import { indexDocumentURL, updateDocumentURL } from "../../url/indexing";
import { recommendURL } from "../../url/recommend";
import { recommendFeedbackURL } from "../../url/recommend-feedback";
import { searchURL } from "../../url/search";
import { searchFeedbackURL } from "../../url/search-feedback";
import { checkExistence } from "../../url/url.utils";
import { URLBuilderOptions } from "./options";



/**
 * 
 * each function is bound with execution context which is defined here:
 */
export interface URLBuilderExecutionContext {
    options: URLBuilderOptions,
    checkExistence: typeof checkExistence,
    indexDocumentURL: typeof indexDocumentURL,
    updateDocumentURL: typeof updateDocumentURL,
    searchURL: typeof searchURL,
    searchFeedbackURL: typeof searchFeedbackURL,
    autocompleteURL: typeof autocompleteURL,
    autocompleteFeedbackURL: typeof autocompleteFeedbackURL,
    feedbackURL: typeof recommendURL,
    recommendFeedbackURL: typeof recommendFeedbackURL,
    deleteDocumentURL: typeof deleteDocumentURL,
}
