// @lableb/javascript-sdk/src/url/recommend-feedback.ts
import { camelCaseToSnackCase, customIdentity } from "../helpers";
import { pickBy } from "../helpers/pick-by";
import { RecommendFeedbackURLOptions, URLBuilderExecutionContext } from "../types/url";
import { buildURLParams } from "./url.utils";



/**
 * Generate Recommendation feed back URL
 * 
 * @param {RecommendFeedbackOptions} recommendFeedbackOptions 
 * @param {number} recommendFeedbackOptions.sourceId - recommendation input document id (that was responsible for generating the recommendations)
 * @param {string} recommendFeedbackOptions.sourceTitle - recommendation input document title
 * @param {string} recommendFeedbackOptions.sourceUrl - recommendation input document url
 * @param {number} recommendFeedbackOptions.targetId - user chosen recommendation item id (that was returned from the recommendation response)
 * @param {string} recommendFeedbackOptions.targetTitle - user chosen recommendation item title
 * @param {string} recommendFeedbackOptions.targetUrl - user chosen recommendation item url
 * @param {number} recommendFeedbackOptions.itemOrder - user chosen recommendation item order in recommendation response
 * @param {string} recommendFeedbackOptions.sessionId - unique session id per user
 * @param {number} recommendFeedbackOptions.userId - user id
 * @param {string} recommendFeedbackOptions.userIp - user ip address
 * @param {string} recommendFeedbackOptions.userCountry - iso 3166 country code
 *
 * @return {string} - URL used for the Feedback Recommend Document request
 */
export function recommendFeedbackURL(

    this: URLBuilderExecutionContext,

    recommendFeedbackOptions: RecommendFeedbackURLOptions = {}

) {

    this.checkExistence({
        functionName: 'recommendFeedbackURL',
        properties: ["projectName", "collectionName", "recommendHandler"]
    });

    const { baseURL, projectName, collectionName, recommendHandler, } = this.options;

    /**
     * convert to snack case and remove undefined values
     */
    let urlParams = buildURLParams(
        pickBy(
            camelCaseToSnackCase({
                ...recommendFeedbackOptions,
                token: this.options.searchToken,
            }),
            customIdentity
        )
    );

    urlParams = urlParams ? `?${urlParams}` : ``;


    return `${baseURL}/projects/${projectName}/collections/${collectionName}/recommend/${recommendHandler}/feedback/hits${urlParams}`;
}
