// @lableb/javascript-sdk/src/core/search-feedback.ts
import { LablebSDKContext } from ".";
import { SearchFeedbackURLOptions } from "../types/url";


/**
 * Send Search feedback to Lableb Cloud
 * 
 * @param {object} options - options object 
 * @param {string} options.query - query sent with the search request
 * @param {number} options.itemId - document id from search response(default to `id` but might be changed in project configuration so you might see it under different name)
 * @param {number} options.itemOrder - the order the item came in from the response
 * @param {string} options.url - the item url on the host website
 * @param {string} options.sessionId - unique session id per user
 * @param {number} options.userId - unique user id
 * @param {string} options.userIp - user ip address
 * @param {string} options.userCountry - user country iso 3166 code
 * 
 * @returns {Promise} - feedback response
 */
export function searchFeedback(

    this: LablebSDKContext,

    options: Omit<SearchFeedbackURLOptions, "sessionId"> = { query: '' }

) {

    return this.httpClient.get({
        url: this.urlBuilder.searchFeedbackURL({
            ...options,
            sessionId: this.sessionId,
        }),
    });
}





/**
 * Directly send search feedback from url
 * @param {string} url 
 * 
 * @returns {Promise} - search feedback response
 */
export function searchFeedbackFromURL(

    this: LablebSDKContext,

    url: string

) {

    return this.httpClient.get({
        url
    });
}
